storage/range.go

```diff
@@ -760,57 +747,103 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
 // struct to get args and reply and then applying the command to the
 // state machine via applyRaftCommand(). The error result is sent on
 // the command's done channel, if available.
-//
-// TODO(spencer): Differentiate between errors caused by the normal culprits --
-// bad inputs from clients, stale information, etc. and errors which might
-// cause the range replicas to diverge -- running out of disk space, underlying
-// rocksdb corruption, etc. Do a careful code audit to make sure we identify
-// errors which should be classified as a ReplicaCorruptionError--when those
-// bubble up to the point where we've just tried to execute a Raft command, the
-// Raft replica would need to stall itself.
 func (r *Range) processRaftCommand(idKey cmdIDKey, index uint64, raftCmd proto.InternalRaftCommand) error {
 	if index == 0 {
 		log.Fatalc(r.context(), "processRaftCommand requires a non-zero index")
 	}
+
 	r.Lock()
 	cmd := r.pendingCmds[idKey]
 	delete(r.pendingCmds, idKey)
 	r.Unlock()
 
 	args := raftCmd.Cmd.GetValue().(proto.Request)
+	var reply proto.Response
+	var ctx context.Context
 	if cmd != nil {
 		// We initiated this command, so use the caller-supplied reply.
-		err := r.applyRaftCommand(cmd.ctx, index, proto.RaftNodeID(raftCmd.OriginNodeID), args, cmd.Reply)
-		cmd.done <- err
-		return err
+		reply = cmd.Reply
+		ctx = cmd.ctx
+	} else {
+		// This command originated elsewhere so we must create a new reply buffer.
+		reply = args.CreateReply()
+		ctx = r.context()
+	}
+
+	var err error
+	// Verify the leader lease is held; Note that we don't require the leader
+	// lease when trying to grant the leader lease!
+	lease := r.getLease()
+	originNode := proto.RaftNodeID(raftCmd.OriginNodeID)
+	var intents []proto.Intent
+	if args.Method() != proto.InternalLeaderLease &&
+		(!lease.OwnedBy(originNode) || !lease.Covers(args.Header().Timestamp)) {
+		// This error case is somewhat special: Any command that makes it into
```

Comment:
The comment  is so big for a relatively straightforward situation. How about we rephrase as:

    // If the lease was lost by this replica during the time the command
    // took from Raft proposal until now, redirect to the new lease holder.
    // This is strictly necessary because the new lease holder may have
    // already served a read which conflicts with this write operation.